materialized value of Flow.lazyInit must be a Future #24670

swachter opened the pull request:
I fixed issue #24670 by changing the materialized value of LazyFlow from `M` to `Future[Option[M]]`.

There may be one issue left: I removed the support for a supervision strategy. It was not clear to me how supervision should work in case that the lazy flow has to be materialized again. I must admit that I am not very familiar with supervision of flows.

Commit:
change materialized value of LazyFlow from [M] to Future[Option[M]]

Commit:
remove whitespace

Commit:
improve docu

ktoso commented:
OK TO TEST

ktoso commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -523,8 +514,8 @@ object Flow {
    */
-  def lazyInit[I, O, M](flowFactory: I ⇒ Future[Flow[I, O, M]], fallback: () ⇒ M): Flow[I, O, M] =
-    Flow.fromGraph(new LazyFlow[I, O, M](flowFactory, fallback))
+  def lazyInit[I, O, M](flowFactory: I ⇒ Future[Flow[I, O, M]]): Flow[I, O, Future[Option[M]]] =
```
If we change the signature like this then please do keep the old method and mark it as deprecated, this way we can phase it out without breaking binary compatibility for users

swachter replied:
Ok. I  can add a method implementing the old signature. The implementation does always use the value created by the `fallback` function as the materialized value. (There is no other way to implement that signature.)

ktoso replied:
Yeah, it will have to be such crippled impl

swachter replied:
There is a slight complication: method overloading does not work well with function parameter derivation, i.e. the compiler does not infer that the `flowFactory` takes an `I`. The parameter type has to specified manually: `Flow.lazyInit[Int, Int, Mat]((_: Int) =>...)` 

I wonder if it is really that useful that the `flowFactory` is a function that gets the first element as input. In addition, it would be more convenient (type derivation wise) if a call by name argument or a function without parameters is used:

```
def lazyInit[In, Out, Mat](flowFactory: => Future[Flow[In, Out, Mat]]]): ...
```

This would allow to call `Flow.lazyInit(someFlow)` without having to specify any types.

swachter replied:
One more note: the mentioned overloading issue would hit all usage sites of the deprecated method if the new method has the same name. Maybe the new method should get a new name, too?

swachter replied:
Yet another note: `Source` distinguishes `Source.lazily` and `Source.lazilyAsync` the latter returning a `Future` from the factory. Therefore a suitable name for the new method might be `Flow.lazyInitAsync`. After its deprecation phase the original method may be replaced by a new implementation `def lazyInit[In, Out, Mat](flowFactory: () => Flow[In, Out, Mat]`): ...`

Commit:
restore old Flow.lazyInit method and add new Flow.lazyInitAsync method

Commit:
fix deprecation messages

swachter commented:
I see that the build fails because of mima reporting an incompatiblity. Unfortunately I do not see the exact reason. How can one access the mima report?

The public API of Flow.lazyInit  (Scala + Java) should be the same again. The @InternalAPI of the LazyFlow class however, has changed. I dropped the fallback parameter. Does this cause the check to fail?

ktoso commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazyFlowSpec.scala:
```diff
@@ -140,44 +135,35 @@ class LazyFlowSpec extends StreamSpec {
+    //      sourceSub.sendNext(1)
+    //      probe.expectNext(1)
+    //      probe.cancel()
+    //    }
```
why? please remove if not valid anymore

swachter replied:
I did not remove these lines yet because I first wanted to discuss the supervision issue (cf. may initial comment - "There may be one issue left: ...").

@gosubpl 
I would like to remove supervision support / logic from LazyFlow:

First: The new implementation `Flow.lazyInitAsync` no more provides the first element to the `flowFactory` function. If the construction of the inner flow fails then this is caused by "external" factors only. If a retry mechanism is needed then it can be implemented on the Future level by using `Future.recoverWith` or `Future.recover`.

Second: If the materialization of the inner flow fails then there is most likely a logic error in the returned flow blueprint. I do not think that supervision strategies are meant to handle these kinds of failures.

swachter commented:
@gosubpl can you please tell me the sbt command for running mima

gosubpl commented:
should be `mimaReportBinaryIssues`

swachter commented:
The mima issue is:

```
[error]  * method this(scala.Function1,scala.Function0)Unit in class akka.stream.impl.fusing.LazyFlow does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.impl.fusing.LazyFlow.this")
```
i.e. it is because the signature of the LazyFlow constructor changed. The class is annotated with `@InternalAPI`. Therefore I had assumed that it may change. From the ScalaDoc of `@InternalApi`:

```
Marks APIs that are considered internal to Akka and may change at any point in time without any warning.
```

johanandren commented:
I'm afraid MiMa does not understand those annotations (yet) so you will have to explicitly add the `ProblemFilter` entry to `akka-stream/src/main/2.4.11.backwards.excludes`

Commit:
add 2.5.11.backwards.excludes because of changed LazyFlow constructor signature

Commit:
merged master

gosubpl commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazyFlowSpec.scala:
```diff
@@ -113,7 +108,7 @@ class LazyFlowSpec extends StreamSpec {
       val flowProbe = Source.fromPublisher(sourceProbe)
-        .via(Flow.lazyInit[Int, Int, NotUsed](_ ⇒ Future.failed(ex), fallback))
+        .via(Flow.lazyInitAsync[Int, Int, NotUsed](() ⇒ Future.failed(ex)))
         .withAttributes(supervisionStrategy(stoppingDecider))
```
If we decide not to do supervision this probably shouldn't be there.

gosubpl commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazyFlowSpec.scala:
```diff
@@ -126,7 +121,7 @@ class LazyFlowSpec extends StreamSpec {
       val probe = Source.fromPublisher(sourceProbe)
-        .via(Flow.lazyInit[Int, Int, NotUsed](_ ⇒ flowF, fallback))
+        .via(Flow.lazyInitAsync[Int, Int, NotUsed](() ⇒ flowF))
         .withAttributes(supervisionStrategy(stoppingDecider))
```
see above.

gosubpl commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
```diff
@@ -1964,148 +1965,181 @@ private[stream] object Collect {
+        // However, isClosed(in) and subOutlet.isClosed may be different. This happens if upstream completes before
+        // the cached element was pushed.
+        def maybeCompleteStage(): Unit = {
+          if (isClosed(in) && subInlet.isClosed && isClosed(out)) {
```
Maybe a test case for the scenario from the comment above?

swachter replied:
I removed the supervision strategy from the tests.

I changed the test case `complete normally when upstream is complete` into a test case that checks `complete normally when upstream completes BEFORE the stage has switched to the inner flow`. The check is implemented by delaying the completion of the flowFactory future.

I added another test `complete normally when upstream completes AFTER the stage has switched to the inner flow` that checks if completion propagates properly after the stage has switched to the inner flow. That test is implemented by delaying the completion until the first element has passed throught the flow.

Commit:
check switching behaviour

gosubpl commented:
@swachter you need to run the tests in sbt for formatter to format the code (at least compile them)

Commit:
apply formatting

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -543,8 +552,8 @@ object Flow {
    */
-  def lazyInit[I, O, M](flowFactory: I ⇒ Future[Flow[I, O, M]], fallback: () ⇒ M): Flow[I, O, M] =
-    Flow.fromGraph(new LazyFlow[I, O, M](flowFactory, fallback))
+  def lazyInitAsync[I, O, M](flowFactory: () ⇒ Future[Flow[I, O, M]]): Flow[I, O, Future[Option[M]]] =
```
Perhaps we should deprecate/rename `Sink.lazyInit` also to make the naming consistent?

swachter replied:
I agree. If Sink.lazyInit is deprecated there should be a corresponding Sink.lazyInitAsync implementation.

@gosubpl  Do you want to take care of that in a separate pull request?

gosubpl replied:
@swachter @patriknw up to you, gentlemen 😉

swachter replied:
Go for it!

gosubpl replied:
Sadly, might not happen this month.

swachter replied:
Ok. Then I will extend this pull request to cover Sink.lazyInit, too.

gosubpl replied:
👍

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -230,10 +244,13 @@ object Flow {
+    import scala.compat.java8.FutureConverters._
+
+    val sflow = scaladsl.Flow.lazyInitAsync(() ⇒ flowFactory.create().toScala.map(_.asScala)(ExecutionContexts.sameThreadExecutionContext))
+      .mapMaterializedValue(fut ⇒ fut.map(_.fold[Optional[M]](Optional.empty())(m ⇒ Optional.of(m)))(ExecutionContexts.sameThreadExecutionContext).toJava)
```
should probably never be null, but wouldn't hurt to use `Optional.ofNullable`

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -207,20 +207,34 @@ object Flow {
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  @deprecated("Use lazyInitAsync instead.", "2.5.12")
```
perhaps include a reason: ", because materialized value of this method is invalid"

gosubpl replied:
👍

gosubpl reviewed (approved):
Looking good, thanks for fixing!

Commit:
improve deprecation message; improve null safety

Commit:
prevent premature stage completion by setting keepGoing

Commit:
deprecate Sink.lazyInit; add Sink.lazyInitAsync

Commit:
Merge remote-tracking branch 'upstream/master' into fix-24670

swachter commented:
I deprecated Sink.lazyInit and added a corresponding Sink.lazyInitAsync method.

In addition, in the implementation of LazyFlow I fixed a subtle bug concerning stage completion and added `setKeepGoing(true)` in order to avoid premature stage completion.

Commit:
apply formatting

swachter commented:
Mima fails. I have no clue what the issue is:

```
[error] java.lang.RuntimeException: akka-stream: Binary compatibility check failed!
[error] 	at scala.sys.package$.error(package.scala:27)
[error] 	at com.typesafe.tools.mima.plugin.SbtMima$.reportModuleErrors(SbtMima.scala:114)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$11(MimaPlugin.scala:78)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$11$adapted(MimaPlugin.scala:61)
[error] 	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:231)
[error] 	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
[error] 	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$9(MimaPlugin.scala:61)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$9$adapted(MimaPlugin.scala:45)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:39)
[error] 	at sbt.std.Transform$$anon$4.work(System.scala:66)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:262)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error] 	at sbt.Execute.work(Execute.scala:271)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:262)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:174)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:36)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] 	at java.lang.Thread.run(Thread.java:748)
[error] (akka-stream / mimaReportBinaryIssues) akka-stream: Binary compatibility check failed!
```

johanandren commented:
You can run only mima locally like so: `akka-stream/mimaReportBinaryIssues` to get only the failing output

Commit:
add ProblemFilter.exclude for changed LazySink.this

swachter commented:
Thanks for the advice! The output complains:

```
[error]  * method this(scala.Function1,scala.Function0)Unit in class akka.stream.impl.LazySink does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.impl.LazySink.this")
```
I will add that filter.

johanandren commented:
Yes, that is safe as it is internal.

Commit:
Merge remote-tracking branch 'upstream/master' into fix-24670

Commit:
Merge remote-tracking branch 'upstream/master' into fix-24670

ktoso committed:
Merge branch 'master' into fix-24670

ktoso committed:
Update Sink.scala

ktoso committed:
Update Sink.scala

ktoso committed:
Update Flow.scala

ktoso committed:
Update Flow.scala

ktoso commented:
LGTM, added missing Java `@Deprecated` and will add missing entry in stages-overview

ktoso commented on akka-stream/src/main/scala/akka/stream/impl/Sinks.scala:
```diff
@@ -456,110 +449,130 @@ import scala.collection.generic.CanBuildFrom
+            case Success(sink) ⇒
+              // check if the stage is still in need for the lazy sink
+              // (there could have been an onUpstreamFailure in the meantime that has completed the promise)
+              if (!promise.isCompleted) {
```
FYI not a good pattern to guard like this; in some cases can be racy between the isCompleted check and the success call. Usually `trySuccess` is the way to go. In this case it'll be fine since only one trigger for `cb` and concurrency guarded due to the stage and callback, but tip in general

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazySinkSpec.scala:
```diff
@@ -86,22 +85,21 @@ class LazySinkSpec extends StreamSpec {
       sourceSub.expectRequest(1)
       sourceSub.sendNext(0)
-      val probe = Await.result(futureProbe, 300.millis)
+      val probe = Await.result(futureProbe, 300.millis).get
```
I know this was not introduced by you, but perhaps you can follow up in another PR.
No reason for using such tight timeouts here (and several other places), use `remainingOrDefault` instead.
see https://github.com/akka/akka/blob/master/CONTRIBUTING.md#preferred-ways-to-use-timeouts-in-tests

ktoso replied:
Addressing this in follow up as well then

ktoso commented:
Thanks a ton @swachter ! :-)

patriknw commented:
update of stages-overview.md was missing, perhaps you can follow up in another PR?

ktoso commented:
Yes I'm now changing just that actually, easier in a follow up than adding changed into this PR

ktoso commented:
PRed
